### ISSUE TITLE
MODIS and SEADAS test failures

### ISSUE THREAD

**Issue by gerritholl** [pytroll/satpy#2884]
**Describe the bug**

MODIS and SEADAS tests are failing with latest versions of all libraries.

**To Reproduce**

```bash
mamba update --all
pytest reader_tests/modis_tests/ reader_tests/test_seadas_l2.py
```
**Expected behavior**

I expect all tests to pass.

**Actual results**

25 test failures:

```
================================================================== short test summary info ===================================================================
FAILED modis_tests/test_modis_l1b.py::TestModisL1b::test_load_longitude_latitude[modis_l1b_nasa_mod021km_file-True-False-False-1000] - pyhdf.error.HDF4Error: info (59): Invalid arguments to routine
FAILED modis_tests/test_modis_l1b.py::TestModisL1b::test_load_longitude_latitude[modis_l1b_imapp_1000m_file-True-False-False-1000] - pyhdf.error.HDF4Error: info (59): Invalid arguments to routine
FAILED modis_tests/test_modis_l1b.py::TestModisL1b::test_load_longitude_latitude[modis_l1b_nasa_mod02hkm_file-False-True-True-250] - pyhdf.error.HDF4Error: info (59): Invalid arguments to routine
FAILED modis_tests/test_modis_l1b.py::TestModisL1b::test_load_longitude_latitude[modis_l1b_nasa_mod02qkm_file-False-True-True-250] - pyhdf.error.HDF4Error: info (59): Invalid arguments to routine
FAILED modis_tests/test_modis_l1b.py::TestModisL1b::test_load_longitude_latitude[modis_l1b_nasa_1km_mod03_files-True-True-True-250] - pyhdf.error.HDF4Error: info (59): Invalid arguments to routine
FAILED modis_tests/test_modis_l1b.py::TestModisL1b::test_load_sat_zenith_angle - pyhdf.error.HDF4Error: info (59): Invalid arguments to routine
FAILED modis_tests/test_modis_l1b.py::TestModisL1b::test_load_vis - pyhdf.error.HDF4Error: info (59): Invalid arguments to routine
FAILED modis_tests/test_modis_l1b.py::TestModisL1b::test_load_vis_saturation[False] - pyhdf.error.HDF4Error: info (59): Invalid arguments to routine
FAILED modis_tests/test_modis_l1b.py::TestModisL1b::test_load_vis_saturation[True] - pyhdf.error.HDF4Error: info (59): Invalid arguments to routine
FAILED modis_tests/test_modis_l2.py::TestModisL2::test_load_longitude_latitude[modis_l2_nasa_mod35_file-True-False-False-1000] - pyhdf.error.HDF4Error: info (59): Invalid arguments to routine
FAILED modis_tests/test_modis_l2.py::TestModisL2::test_load_quality_assurance - pyhdf.error.HDF4Error: info (59): Invalid arguments to routine
FAILED modis_tests/test_modis_l2.py::TestModisL2::test_load_category_dataset[modis_l2_nasa_mod35_mod03_files-loadables0-1000-1000-True] - pyhdf.error.HDF4Error: info (59): Invalid arguments to routine
FAILED modis_tests/test_modis_l2.py::TestModisL2::test_load_category_dataset[modis_l2_imapp_mask_byte1_geo_files-loadables1-None-1000-True] - pyhdf.error.HDF4Error: info (59): Invalid arguments to routine
FAILED modis_tests/test_modis_l2.py::TestModisL2::test_load_250m_cloud_mask_dataset[modis_l2_nasa_mod35_file-False] - pyhdf.error.HDF4Error: info (59): Invalid arguments to routine
FAILED modis_tests/test_modis_l2.py::TestModisL2::test_load_250m_cloud_mask_dataset[modis_l2_nasa_mod35_mod03_files-True] - pyhdf.error.HDF4Error: info (59): Invalid arguments to routine
FAILED modis_tests/test_modis_l2.py::TestModisL2::test_load_l2_dataset[modis_l2_nasa_mod06_file-loadables0-5000-True-4.0] - pyhdf.error.HDF4Error: info (59): Invalid arguments to routine
FAILED modis_tests/test_modis_l2.py::TestModisL2::test_load_l2_dataset[modis_l2_imapp_snowmask_file-loadables1-1000-False-1.0] - pyhdf.error.HDF4Error: info (59): Invalid arguments to routine
FAILED modis_tests/test_modis_l2.py::TestModisL2::test_load_l2_dataset[modis_l2_imapp_snowmask_geo_files-loadables2-1000-True-1.0] - pyhdf.error.HDF4Error: info (59): Invalid arguments to routine
FAILED modis_tests/test_modis_l3.py::TestModisL3::test_load_l3_dataset - pyhdf.error.HDF4Error: info (59): Invalid arguments to routine
FAILED test_seadas_l2.py::TestSEADAS::test_load_chlor_a[False-seadas_l2_modis_chlor_a-Aqua-exp_sensor0-10] - pyhdf.error.HDF4Error: info (59): Invalid arguments to routine
FAILED test_seadas_l2.py::TestSEADAS::test_load_chlor_a[False-seadas_l2_viirs_npp_chlor_a-Suomi-NPP-exp_sensor1-16] - pyhdf.error.HDF4Error: info (59): Invalid arguments to routine
FAILED test_seadas_l2.py::TestSEADAS::test_load_chlor_a[False-seadas_l2_viirs_j01_chlor_a-NOAA-20-exp_sensor2-16] - pyhdf.error.HDF4Error: info (59): Invalid arguments to routine
FAILED test_seadas_l2.py::TestSEADAS::test_load_chlor_a[True-seadas_l2_modis_chlor_a-Aqua-exp_sensor0-10] - pyhdf.error.HDF4Error: info (59): Invalid arguments to routine
FAILED test_seadas_l2.py::TestSEADAS::test_load_chlor_a[True-seadas_l2_viirs_npp_chlor_a-Suomi-NPP-exp_sensor1-16] - pyhdf.error.HDF4Error: info (59): Invalid arguments to routine
FAILED test_seadas_l2.py::TestSEADAS::test_load_chlor_a[True-seadas_l2_viirs_j01_chlor_a-NOAA-20-exp_sensor2-16] - pyhdf.error.HDF4Error: info (59): Invalid arguments to routine
========================================================= 25 failed, 16 passed, 1 warning in 14.61s ==========================================================
```

**Environment Info:**
 - OS: openSUSE Leap 15.3
 - Satpy Version: 0.51.0
 - Readers and writers dependencies (when relevant):

Extract:

```
modis_l1b:  ok
modis_l2:  ok
modis_l3:  ok
seadas_l2:  ok
```

**Additional context**

Must be some recent dependency that got upgraded, as it was working before I did `mamba update --all`.  The updates were:

```
     aiohttp  {3.9.5 (conda-forge/linux-64) -> 3.10.5 (conda-forge/linux-64)}                                                                                                                                                                                                                                        [67/216]
     attrs  {23.2.0 (conda-forge/noarch) -> 24.2.0 (conda-forge/noarch)}
     aws-c-auth  {0.7.22 (conda-forge/linux-64) -> 0.7.25 (conda-forge/linux-64)}
     aws-c-cal  {0.6.15 (conda-forge/linux-64) -> 0.7.3 (conda-forge/linux-64)}
     aws-c-common  {0.9.23 (conda-forge/linux-64) -> 0.9.27 (conda-forge/linux-64)}
     aws-c-compression  {0.2.18 (conda-forge/linux-64) -> 0.2.18 (conda-forge/linux-64)}
     aws-c-event-stream  {0.4.2 (conda-forge/linux-64) -> 0.4.3 (conda-forge/linux-64)}
     aws-c-http  {0.8.2 (conda-forge/linux-64) -> 0.8.7 (conda-forge/linux-64)}
     aws-c-io  {0.14.9 (conda-forge/linux-64) -> 0.14.18 (conda-forge/linux-64)}
     aws-c-mqtt  {0.10.4 (conda-forge/linux-64) -> 0.10.4 (conda-forge/linux-64)}
     aws-c-s3  {0.5.10 (conda-forge/linux-64) -> 0.6.4 (conda-forge/linux-64)}
     aws-c-sdkutils  {0.1.16 (conda-forge/linux-64) -> 0.1.19 (conda-forge/linux-64)}
     aws-checksums  {0.1.18 (conda-forge/linux-64) -> 0.1.18 (conda-forge/linux-64)}
     aws-crt-cpp  {0.26.12 (conda-forge/linux-64) -> 0.27.5 (conda-forge/linux-64)}
     aws-sdk-cpp  {1.11.329 (conda-forge/linux-64) -> 1.11.379 (conda-forge/linux-64)}
     azure-core-cpp  {1.12.0 (conda-forge/linux-64) -> 1.13.0 (conda-forge/linux-64)}
     azure-identity-cpp  {1.8.0 (conda-forge/linux-64) -> 1.8.0 (conda-forge/linux-64)}
     azure-storage-blobs-cpp  {12.11.0 (conda-forge/linux-64) -> 12.12.0 (conda-forge/linux-64)}
     azure-storage-common-cpp  {12.6.0 (conda-forge/linux-64) -> 12.7.0 (conda-forge/linux-64)}
     azure-storage-files-datalake-cpp  {12.10.0 (conda-forge/linux-64) -> 12.11.0 (conda-forge/linux-64)}
     backports  {1.0 (conda-forge/noarch) -> 1.0 (conda-forge/noarch)}
     bokeh  {3.4.2 (conda-forge/noarch) -> 3.5.1 (conda-forge/noarch)}
     boto3  {1.34.138 (conda-forge/noarch) -> 1.35.1 (conda-forge/noarch)}
     botocore  {1.34.138 (conda-forge/noarch) -> 1.35.2 (conda-forge/noarch)}
     bzip2  {1.0.8 (conda-forge/linux-64) -> 1.0.8 (conda-forge/linux-64)}
     c-ares  {1.28.1 (conda-forge/linux-64) -> 1.33.0 (conda-forge/linux-64)}
     cffi  {1.16.0 (conda-forge/linux-64) -> 1.17.0 (conda-forge/linux-64)}
     coverage  {7.5.4 (conda-forge/linux-64) -> 7.6.1 (conda-forge/linux-64)}
     curl  {8.8.0 (conda-forge/linux-64) -> 8.9.1 (conda-forge/linux-64)}
     dask  {2024.6.2 (conda-forge/noarch) -> 2024.8.1 (conda-forge/noarch)}
     dask-core  {2024.6.2 (conda-forge/noarch) -> 2024.8.1 (conda-forge/noarch)}
     dask-expr  {1.1.6 (conda-forge/noarch) -> 1.1.11 (conda-forge/noarch)}
     distributed  {2024.6.2 (conda-forge/noarch) -> 2024.8.1 (conda-forge/noarch)}
     exceptiongroup  {1.2.0 (conda-forge/noarch) -> 1.2.2 (conda-forge/noarch)}
     flake8  {7.1.0 (conda-forge/noarch) -> 7.1.1 (conda-forge/noarch)}
     fmt  {10.2.1 (conda-forge/linux-64) -> 11.0.2 (conda-forge/linux-64)}
     fonttools  {4.53.0 (conda-forge/linux-64) -> 4.53.1 (conda-forge/linux-64)}
     geos  {3.12.1 (conda-forge/linux-64) -> 3.12.2 (conda-forge/linux-64)}
     geotiff  {1.7.3 (conda-forge/linux-64) -> 1.7.3 (conda-forge/linux-64)}
     gettext  {0.22.5 (conda-forge/linux-64) -> 0.22.5 (conda-forge/linux-64)}
     gettext-tools  {0.22.5 (conda-forge/linux-64) -> 0.22.5 (conda-forge/linux-64)}
     git  {2.45.2 (conda-forge/linux-64) -> 2.46.0 (conda-forge/linux-64)}
     glib  {2.80.2 (conda-forge/linux-64) -> 2.80.3 (conda-forge/linux-64)}
     glib-tools  {2.80.2 (conda-forge/linux-64) -> 2.80.3 (conda-forge/linux-64)}
     graphviz  {11.0.0 (conda-forge/linux-64) -> 12.0.0 (conda-forge/linux-64)}
     gst-plugins-base  {1.24.5 (conda-forge/linux-64) -> 1.24.6 (conda-forge/linux-64)}
     gstreamer  {1.24.5 (conda-forge/linux-64) -> 1.24.6 (conda-forge/linux-64)}
     gtk2  {2.24.33 (conda-forge/linux-64) -> 2.24.33 (conda-forge/linux-64)}
     harfbuzz  {8.5.0 (conda-forge/linux-64) -> 9.0.0 (conda-forge/linux-64)}
     identify  {2.5.36 (conda-forge/noarch) -> 2.6.0 (conda-forge/noarch)}
     importlib-metadata  {8.0.0 (conda-forge/noarch) -> 8.4.0 (conda-forge/noarch)}
     importlib_metadata  {8.0.0 (conda-forge/noarch) -> 8.4.0 (conda-forge/noarch)}
     json-c  {0.17 (conda-forge/linux-64) -> 0.17 (conda-forge/linux-64)}
     libabseil  {20240116.2 (conda-forge/linux-64) -> 20240116.2 (conda-forge/linux-64)}
     libarrow  {16.1.0 (conda-forge/linux-64) -> 17.0.0 (conda-forge/linux-64)}
     libarrow-acero  {16.1.0 (conda-forge/linux-64) -> 17.0.0 (conda-forge/linux-64)}
     libarrow-dataset  {16.1.0 (conda-forge/linux-64) -> 17.0.0 (conda-forge/linux-64)}
     libarrow-flight  {16.1.0 (conda-forge/linux-64) -> 17.0.0 (conda-forge/linux-64)}
     libarrow-flight-sql  {16.1.0 (conda-forge/linux-64) -> 17.0.0 (conda-forge/linux-64)}
     libarrow-gandiva  {16.1.0 (conda-forge/linux-64) -> 17.0.0 (conda-forge/linux-64)}
     libarrow-substrait  {16.1.0 (conda-forge/linux-64) -> 17.0.0 (conda-forge/linux-64)}
     libasprintf  {0.22.5 (conda-forge/linux-64) -> 0.22.5 (conda-forge/linux-64)}
     libasprintf-devel  {0.22.5 (conda-forge/linux-64) -> 0.22.5 (conda-forge/linux-64)}
     libblas  {3.9.0 (conda-forge/linux-64) -> 3.9.0 (conda-forge/linux-64)}
     libboost-headers  {1.85.0 (conda-forge/linux-64) -> 1.86.0 (conda-forge/linux-64)}
     libcblas  {3.9.0 (conda-forge/linux-64) -> 3.9.0 (conda-forge/linux-64)}
     libclang  {18.1.8 (conda-forge/linux-64) -> 18.1.8 (conda-forge/linux-64)}
     libclang13  {18.1.8 (conda-forge/linux-64) -> 18.1.8 (conda-forge/linux-64)}
     libcurl  {8.8.0 (conda-forge/linux-64) -> 8.9.1 (conda-forge/linux-64)}
     libdeflate  {1.20 (conda-forge/linux-64) -> 1.21 (conda-forge/linux-64)}
     libgcrypt  {1.11.0 (conda-forge/linux-64) -> 1.11.0 (conda-forge/linux-64)}
     libgdal  {3.9.1 (conda-forge/linux-64) -> 3.9.2 (conda-forge/linux-64)}
     libgettextpo  {0.22.5 (conda-forge/linux-64) -> 0.22.5 (conda-forge/linux-64)}
     libgettextpo-devel  {0.22.5 (conda-forge/linux-64) -> 0.22.5 (conda-forge/linux-64)}
     libglib  {2.80.2 (conda-forge/linux-64) -> 2.80.3 (conda-forge/linux-64)}
     libgoogle-cloud  {2.25.0 (conda-forge/linux-64) -> 2.28.0 (conda-forge/linux-64)}
     libgoogle-cloud-storage  {2.25.0 (conda-forge/linux-64) -> 2.28.0 (conda-forge/linux-64)}
     libkml  {1.3.0 (conda-forge/linux-64) -> 1.3.0 (conda-forge/linux-64)}
     liblapack  {3.9.0 (conda-forge/linux-64) -> 3.9.0 (conda-forge/linux-64)}
     libllvm18  {18.1.8 (conda-forge/linux-64) -> 18.1.8 (conda-forge/linux-64)}
     libnl  {3.9.0 (conda-forge/linux-64) -> 3.10.0 (conda-forge/linux-64)}
     libopenblas  {0.3.27 (conda-forge/linux-64) -> 0.3.27 (conda-forge/linux-64)}
     libparquet  {16.1.0 (conda-forge/linux-64) -> 17.0.0 (conda-forge/linux-64)}
     libpq  {16.3 (conda-forge/linux-64) -> 16.4 (conda-forge/linux-64)}
     librsvg  {2.58.1 (conda-forge/linux-64) -> 2.58.3 (conda-forge/linux-64)}
     librttopo  {1.1.0 (conda-forge/linux-64) -> 1.1.0 (conda-forge/linux-64)}
     libspatialite  {5.1.0 (conda-forge/linux-64) -> 5.1.0 (conda-forge/linux-64)}
     libthrift  {0.19.0 (conda-forge/linux-64) -> 0.20.0 (conda-forge/linux-64)}
     libtiff  {4.6.0 (conda-forge/linux-64) -> 4.6.0 (conda-forge/linux-64)}
     libxml2  {2.12.7 (conda-forge/linux-64) -> 2.12.7 (conda-forge/linux-64)}
     matplotlib  {3.9.0 (conda-forge/linux-64) -> 3.9.1 (conda-forge/linux-64)}
     matplotlib-base  {3.9.0 (conda-forge/linux-64) -> 3.9.1 (conda-forge/linux-64)}
     mysql-common  {8.3.0 (conda-forge/linux-64) -> 8.3.0 (conda-forge/linux-64)}
     mysql-libs  {8.3.0 (conda-forge/linux-64) -> 8.3.0 (conda-forge/linux-64)}
     nss  {3.101 (conda-forge/linux-64) -> 3.103 (conda-forge/linux-64)}
     numcodecs  {0.12.1 (conda-forge/linux-64) -> 0.13.0 (conda-forge/linux-64)}
     numpy  {2.0.0 (conda-forge/linux-64) -> 2.0.1 (conda-forge/linux-64)}
     orc  {2.0.1 (conda-forge/linux-64) -> 2.0.2 (conda-forge/linux-64)}
     pango  {1.54.0 (conda-forge/linux-64) -> 1.54.0 (conda-forge/linux-64)}
     pcre2  {10.44 (conda-forge/linux-64) -> 10.44 (conda-forge/linux-64)}
     pip  {24.0 (conda-forge/noarch) -> 24.2 (conda-forge/noarch)}
     poppler  {24.04.0 (conda-forge/linux-64) -> 24.08.0 (conda-forge/linux-64)}
     postgresql  {16.3 (conda-forge/linux-64) -> 16.4 (conda-forge/linux-64)}
     pre-commit  {3.7.1 (conda-forge/noarch) -> 3.8.0 (conda-forge/noarch)}
     proj  {9.4.1 (conda-forge/linux-64) -> 9.4.1 (conda-forge/linux-64)}
     pure_eval  {0.2.2 (conda-forge/noarch) -> 0.2.3 (conda-forge/noarch)}
     pyarrow  {16.1.0 (conda-forge/linux-64) -> 17.0.0 (conda-forge/linux-64)}
     pyarrow-core  {16.1.0 (conda-forge/linux-64) -> 17.0.0 (conda-forge/linux-64)}
     pycodestyle  {2.12.0 (conda-forge/noarch) -> 2.12.1 (conda-forge/noarch)}
     pyproj  {3.6.1 (conda-forge/linux-64) -> 3.6.1 (conda-forge/linux-64)}
     pyresample  {1.28.4 (conda-forge/linux-64) -> 1.29.0 (conda-forge/linux-64)}
     pyspectral  {0.13.2 (conda-forge/noarch) -> 0.13.4 (conda-forge/noarch)}
     pytest  {8.2.2 (conda-forge/noarch) -> 8.3.2 (conda-forge/noarch)}
     python_abi  {3.12 (conda-forge/linux-64) -> 3.12 (conda-forge/linux-64)}
     pyyaml  {6.0.1 (conda-forge/linux-64) -> 6.0.2 (conda-forge/linux-64)}
     pyzmq  {26.0.3 (conda-forge/linux-64) -> 26.1.1 (conda-forge/linux-64)}
     qhull  {2020.2 (conda-forge/linux-64) -> 2020.2 (conda-forge/linux-64)}
     qt-main  {5.15.8 (conda-forge/linux-64) -> 5.15.8 (conda-forge/linux-64)}
     rasterio  {1.3.10 (conda-forge/linux-64) -> 1.3.10 (conda-forge/linux-64)}
     rdma-core  {52.0 (conda-forge/linux-64) -> 53.0 (conda-forge/linux-64)}
     rioxarray  {0.15.7 (conda-forge/noarch) -> 0.17.0 (conda-forge/noarch)}
     s2n  {1.4.16 (conda-forge/linux-64) -> 1.5.0 (conda-forge/linux-64)}
     scipy  {1.14.0 (conda-forge/linux-64) -> 1.14.0 (conda-forge/linux-64)}
     setuptools  {70.1.1 (conda-forge/noarch) -> 72.2.0 (conda-forge/noarch)}
     shapely  {2.0.4 (conda-forge/linux-64) -> 2.0.6 (conda-forge/linux-64)}
     snappy  {1.2.0 (conda-forge/linux-64) -> 1.2.1 (conda-forge/linux-64)}
     snuggs  {1.4.7 (conda-forge/noarch) -> 1.4.7 (conda-forge/noarch)}
     spdlog  {1.13.0 (conda-forge/linux-64) -> 1.14.1 (conda-forge/linux-64)}
     sphinx  {7.3.7 (conda-forge/noarch) -> 7.4.7 (conda-forge/noarch)}
     sphinxcontrib-applehelp  {1.0.8 (conda-forge/noarch) -> 2.0.0 (conda-forge/noarch)}
     sphinxcontrib-devhelp  {1.0.6 (conda-forge/noarch) -> 2.0.0 (conda-forge/noarch)}
     sphinxcontrib-htmlhelp  {2.0.5 (conda-forge/noarch) -> 2.1.0 (conda-forge/noarch)}
     sphinxcontrib-qthelp  {1.0.7 (conda-forge/noarch) -> 2.0.0 (conda-forge/noarch)}
     tiledb  {2.24.1 (conda-forge/linux-64) -> 2.25.0 (conda-forge/linux-64)}
     tqdm  {4.66.4 (conda-forge/noarch) -> 4.66.5 (conda-forge/noarch)}
     trollimage  {1.24.0 (conda-forge/linux-64) -> 1.25.0 (conda-forge/linux-64)}
     ucx  {1.16.0 (conda-forge/linux-64) -> 1.17.0 (conda-forge/linux-64)}
     wheel  {0.43.0 (conda-forge/noarch) -> 0.44.0 (conda-forge/noarch)}
     xarray  {2024.6.0 (conda-forge/noarch) -> 2024.7.0 (conda-forge/noarch)}
     zipp  {3.19.2 (conda-forge/noarch) -> 3.20.0 (conda-forge/noarch)}
     zstandard  {0.22.0 (conda-forge/linux-64) -> 0.23.0 (conda-forge/linux-64)}
    +aiohappyeyeballs-2.4.0 (conda-forge/noarch)
    +libgdal-core-3.9.2 (conda-forge/linux-64)
    +libgdal-fits-3.9.2 (conda-forge/linux-64)
    +libgdal-grib-3.9.2 (conda-forge/linux-64)
    +libgdal-hdf4-3.9.2 (conda-forge/linux-64)
    +libgdal-hdf5-3.9.2 (conda-forge/linux-64)
    +libgdal-jp2openjpeg-3.9.2 (conda-forge/linux-64)
    +libgdal-kea-3.9.2 (conda-forge/linux-64)
    +libgdal-netcdf-3.9.2 (conda-forge/linux-64)
    +libgdal-pdf-3.9.2 (conda-forge/linux-64)
    +libgdal-pg-3.9.2 (conda-forge/linux-64)
    +libgdal-postgisraster-3.9.2 (conda-forge/linux-64)
    +libgdal-tiledb-3.9.2 (conda-forge/linux-64)
    +libgdal-xls-3.9.2 (conda-forge/linux-64)
    +libllvm17-17.0.6 (conda-forge/linux-64)
    +xorg-libxxf86vm-1.1.5 (conda-forge/linux-64)
```

**Comment by mraspaud** [comment]
Ok, so I made a minimal example, and investigated this a bit.
```python
import numpy as np
from pyhdf.SD import SD, SDC, SDS
import xarray as xr
import dask.array as da

HTYPE_TO_DTYPE = {
    SDC.INT8: np.int8,
    SDC.UCHAR: np.uint8,
    SDC.CHAR: np.int8,
    SDC.INT32: np.int32,
    SDC.INT16: np.int16,
    SDC.UINT8: np.uint8,
    SDC.UINT16: np.uint16,
    SDC.UINT32: np.uint32,
    SDC.FLOAT32: np.float32,
    SDC.FLOAT64: np.float64,
}

def create_file():
    """Create a test HDF4 file."""
    filename = "test.hdf"
    h = SD(filename, SDC.WRITE | SDC.CREATE | SDC.TRUNC)
    data = np.arange(10. * 100, dtype=np.float32).reshape((10, 100))
    v1 = h.create("ds1_f", SDC.FLOAT32, (10, 100))
    v1[:] = data
    v2 = h.create("ds1_i", SDC.INT16, (10, 100))
    v2[:] = data.astype(np.int16)

    # Add attributes
    h.test_attr_str = "test_string"
    h.test_attr_int = 0
    h.test_attr_float = 1.2

    for d in [v1, v2]:
        d.test_attr_str = "test_string"
        d.test_attr_int = 0
        d.test_attr_float = 1.2

    h.end()
    return filename


def collect_attrs(file_content, name, attrs):
    for key, value in attrs.items():
        value = np.squeeze(value)
        if issubclass(value.dtype.type, (np.bytes_, np.str_)) and not value.shape:
            value = value.item()  # convert to scalar
            if not isinstance(value, str):
                # python 3 - was scalar numpy array of bytes
                # otherwise python 2 - scalar numpy array of 'str'
                value = value.decode()
            file_content["{}/attr/{}".format(name, key)] = value
        elif not value.shape:
            # convert to a scalar
            file_content["{}/attr/{}".format(name, key)] = value.item()
        else:
            file_content["{}/attr/{}".format(name, key)] = value

def collect_data(file_content, name, obj):
    if isinstance(obj, SDS):
        file_content[name] = obj
        info = obj.info()
        file_content[name + "/dtype"] = np.dtype(HTYPE_TO_DTYPE.get(info[3]))
        file_content[name + "/shape"] = info[2] if isinstance(info[2], (int, float)) else tuple(info[2])


def read_file(filename):
    file_content = dict()
    file_handle = SD(filename, SDC.READ)
    collect_attrs(file_content, "", file_handle.attributes())
    for k in file_handle.datasets().keys():
        collect_data(file_content, k, file_handle.select(k))
        var = file_handle.select(k)
        file_content[k] = var
    return file_content

def from_sds(var, *args, **kwargs):
    """Create a dask array from a SD dataset."""
    info = var.info()
    var.__dict__["dtype"] = np.dtype(HTYPE_TO_DTYPE[info[3]])
    shape = info[2]
    var.__dict__["shape"] = shape if isinstance(shape, (tuple, list)) else tuple(shape)
    return da.from_array(var, *args, **kwargs)

def create_xarray_dataarray(file_content, key):
    val = file_content[key]
    if isinstance(val, SDS):

        dask_arr = from_sds(val)
        attrs = val.attributes()
        val = xr.DataArray(dask_arr, dims=("y", "x"),
                            attrs=attrs)
    return val



filename = create_file()
file_content = read_file(filename)
create_xarray_dataarray(file_content, "ds1_f")
create_xarray_dataarray(file_content, "ds1_i")
```

I came to the conclusion that the update of dask from 2024.8.0 to 2024.8.1 broke this. I don't really know why, but the problem appears when the SDS (pyhdf's variable type) is passed to dask. I suspect that passing the patched SDS dataset as an array to dask is problematic, as maybe dask is exploring more of the array attributes than we would like (in particular the `info` method).

I can see a couple of solutions:
1. Wrap the SDS variable into a custom class (or maybe event an xarray BackendArray) that just exposes the minimal amount of methods and attributes.
2. While investigating this, I realised that the netcdf4 engine of xarray can read hdf4 files... eg `xr.open_dataset("MYD02Hkm_A19057_114915_2019057120321.hdf", engine="netcdf4", chunks="auto")` just works. So maybe we could use that and get rid of pyhdf altogether?

**Comment by djhoese** [comment]
I'm still out sick so I haven't looked in to this much, but I wanted to point out that netcdf4 can read HDF4 if and only if the C library was built with HDF4 support. Conda-forge's package is but I'm not sure if that's done everywhere.

**Comment by gerritholl** [comment]
I would support to use (1) as an immediate stopgap and then consider option (5) later.  If we don't have time to work on (5) within the next months or if it turns out there is an unforeseen problem with (5), we could implement (2) for the midterm (possibly conditional upon the dask version).

**Comment by djhoese** [comment]
I'm confused by your first message @mraspaud. The traceback in the CI seems to suggest this is just us (satpy) requesting the `.info()` not dask. Right?

```
 satpy/readers/hdfeos_base.py:421: in get_dataset
    data = self._load_ds_by_name(dataset_name)
satpy/readers/hdfeos_base.py:385: in _load_ds_by_name
    return self.load_dataset(var_names)
satpy/readers/hdfeos_base.py:222: in load_dataset
    attrs=dataset.attributes())
../../../miniconda3/envs/test-environment/lib/python3.11/site-packages/pyhdf/SD.py:2786: in attributes
    natts = self.info()[4]
../../../miniconda3/envs/test-environment/lib/python3.11/site-packages/pyhdf/SD.py:2128: in info
    _checkErr('info', status, "cannot execute")
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

procName = 'info', val = -1, msg = 'cannot execute'

    def _checkErr(procName, val, msg=""):
    
        if val is None or (not isinstance(val, str) and val < 0):
            #_C._HEprint();
            errCode = _C.HEvalue(1)
            if errCode != 0:
                err = "%s (%d): %s" % (procName, errCode, _C.HEstring(errCode))
            else:
                err = "%s : %s" % (procName, msg)
>           raise HDF4Error(err)
E           pyhdf.error.HDF4Error: info (59): Invalid arguments to routine
```

**Comment by djhoese** [comment]
@mraspaud You may have noticed this and mentioned it already, but if you swap the `val.attributes()` and `dask_arr = from_sds(val)` lines in your test code so attributes is first, then the code succeeds for the first `ds1_f` variable loading, but not the second `ds1_i` loading. So something is not getting closed properly by dask it would seem. More debugging to come...

**Comment by djhoese** [comment]
Got it! Or at least I've got it down to the line of code that starts to break things. It's this PR:

https://github.com/dask/dask/pull/11320

The changes inside `base.py` seem to break things. I noticed while creating a wrapper class around the SDS object before passing to `da.from_array` that there was an unexpected access to `.__dict__` at the `mod.dumps` line added by this PR. If I revert the changes (hacking my local install) then suddenly the minimal example code and the satpy test code all passes. Based on the old code my guess is cloudpickle is doing something suspect to `__dict__` or the object being passed to it. I'm still not completely sure yet.
